lib/enumerator.h

```diff
@@ -90,74 +106,115 @@ class Enumerator {
+        while (moveNext()) result.push_back(getCurrent());
+        return result;
+    }
+
+    /// Enumerate all elements and return the count
+    uint64_t count() {
+        uint64_t found = 0;
+        while (this->moveNext()) found++;
+        return found;
+    }
+
+    /// True if the enumerator has at least one element
+    bool any() { return this->moveNext(); }
+
+    /// The only next element; throws if the enumerator does not have exactly 1 element
+    T single() {
+        bool next = moveNext();
+        if (!next) throw std::logic_error("There is no element for `single()'");
+        T result = getCurrent();
+        next = moveNext();
+        if (next) throw std::logic_error("There are multiple elements when calling `single()'");
+        return result;
+    }
+
+    /// The only next element or default value if none exists; throws if the
+    /// enumerator does not have exactly 0 or 1 element
+    T singleOrDefault() {
+        bool next = moveNext();
+        if (!next) return T{};
+        T result = getCurrent();
+        next = moveNext();
+        if (next) throw std::logic_error("There are multiple elements when calling `single()'");
+        return result;
+    }
+
+    /// Next element, or the default value if none exists
+    T nextOrDefault() {
+        bool next = moveNext();
+        if (!next) return T{};
+        return getCurrent();
+    }
+
+    /// Next element; throws if there are no elements
+    T next() {
+        bool next = moveNext();
+        if (!next) throw std::logic_error("There is no element for `next()'");
+        return getCurrent();
+    }
 };
 
 ////////////////////////// implementation ///////////////////////////////////////////
 // the implementation must be in the header file due to the templates
 
-/* We have to be careful never to invoke getCurrent more than once on each element of the input
-   iterators, since it could have side-effects*/
-
-/* A generic iterator returning elements of type T
-   C is the container type */
+/// We have to be careful never to invoke getCurrent more than once on each
+/// element of the input iterators, since it could have side-effects
 
+/// A generic iterator returning elements of type T.
 template <typename Iter>
-class GenericEnumerator : public Enumerator<typename Iter::value_type> {
+class IteratorEnumerator : public Enumerator<typename Iter::value_type> {
  protected:
     Iter begin;
     Iter end;
     Iter current;
-    cstring name;
+    const char *name;
     friend class Enumerator<typename Iter::value_type>;
 
-    GenericEnumerator(Iter begin, Iter end, cstring name)
+ public:
+    IteratorEnumerator(Iter begin, Iter end, const char *name)
         : Enumerator<typename Iter::value_type>(),
           begin(begin),
           end(end),
           current(begin),
           name(name) {}
 
- public:
-    cstring toString() const { return this->name + ":" + this->stateName(); }
+    [[nodiscard]] std::string toString() const {
```

Comment:
Hmm most other `toString` functions return a cstring. I am not certain whether we should refactor this spotwise or do it all at once it one single PR.

Reply:
Well, I'd go for small iterative changes. Otherwise this (`cstring` change) would end with something of enormous size that is impossible to review / digest

Reply:
Yes. I'm planning to do first sweep one day to get rid of most obvious cases (e.g. we do not need to cstring of format strings or jsons). Then we can discuss the proper API (cstring vs `std::string` vs `std::string_view` usage)